js/private/js_run_devserver.mjs

```diff
@@ -149,17 +258,47 @@ async function syncRecursive(src, dst, writePerm) {
 
 // Sync list of files to the sandbox
 async function sync(files, sandbox, writePerm) {
-    console.error('Syncing...')
+    console.error(`Syncing ${files.length} files && folders...`)
     const startTime = perf_hooks.performance.now()
-    const totalSynced = (
+
+    const [virtualStore1pFiles, remainingFiles] = partitionArray(files, (f) =>
+        is1pVirtualStoreDep(f)
+    )
+
+    if (virtualStore1pFiles.length > 0 && process.env.JS_BINARY__LOG_DEBUG) {
+        console.error(
+            `Syncing ${virtualStore1pFiles.length} first party virtual store dep(s)`
+        )
+    }
+
+    // Sync first-party virtual store files first since correctly syncing direct 1p node_modules
+    // symlinks depends on checking if the virtual store synced files exist.
+    let totalSynced = (
         await Promise.all(
-            files.map(async (file) => {
+            virtualStore1pFiles.map(async (file) => {
                 const src = path.join(RUNFILES_ROOT, file)
                 const dst = path.join(sandbox, file)
-                return await syncRecursive(src, dst, writePerm)
+                return await syncRecursive(src, dst, sandbox, writePerm)
             })
         )
     ).reduce((s, t) => s + t, 0)
+
+    if (remainingFiles.length > 0 && process.env.JS_BINARY__LOG_DEBUG) {
+        console.error(
+            `Syncing ${remainingFiles.length} other files && folders...`
+        )
+    }
+
+    totalSynced += (
+        await Promise.all(
+            remainingFiles.map(async (file) => {
```

Comment:
How is this loop different then the other? Do we need the logging between the two or can we just do `virtualStore1pFiles.concat(remainingFiles).map(...)`?

Reply:
Intentional. Comment explains why. All the 1p files have to complete before any of the remaining file can start to sync so they can't all be queued up in parallel on the same await.